llvm/lib/Transforms/InstCombine/InstructionCombining.cpp

```diff
@@ -5042,10 +5042,29 @@ InstCombinerImpl::pushFreezeToPreventPoisonFromPropagating(FreezeInst &OrigFI) {
   //   Op1.fr = Freeze(Op1)
   //   ... = Inst(Op1.fr, NonPoisonOps...)
 
-  auto CanPushFreeze = [](Value *V) {
-    if (!isa<Instruction>(V) || isa<PHINode>(V))
+  auto CanPushFreeze = [this](Value *V) {
+    if (!isa<Instruction>(V))
       return false;
 
+    if (auto *PN = dyn_cast<PHINode>(V)) {
+      BasicBlock *BB = PN->getParent();
+      SmallPtrSet<BasicBlock *, 8> VisitedBBs;
+      for (Use &U : PN->incoming_values()) {
+        BasicBlock *InBB = PN->getIncomingBlock(U);
+        // We can't move freeze if the start value is the result of a
+        // terminator (e.g. an invoke).
+        if (auto *OpI = dyn_cast<Instruction>(U)) {
+          if (OpI->isTerminator())
+            return false;
+        }
+
+        if (DT.dominates(BB, InBB) || isBackEdge(InBB, BB) ||
+            VisitedBBs.contains(InBB) || match(U.get(), m_Undef()))
```

Comment:
Use VisitedBBs.insert(InBB).second to avoid double HT lookup.

Also this deserves a comment. Ideally we would not have this limitation and make sure multi-edges use the same value.